packages/swr-devtools/src/createSWRDevTools.ts

```diff
@@ -0,0 +1,245 @@
+// We cannot import React nad use hooks that SWR provides because this runs on the application side,
+// we have to use the same React instance with the application
+import { Middleware, Cache, unstable_serialize } from "swr";
+
+import { injectSWRCache } from "./swr-cache";
+
+type EventListener = (...args: any[]) => void;
+
+export class EventEmitter {
+  listeners: EventListener[] = [];
+  subscribe(fn: EventListener) {
+    this.listeners.push(fn);
+    return () => {
+      const index = this.listeners.indexOf(fn);
+      this.listeners.splice(index, 1);
+    };
+  }
+  emit(...args: any[]) {
+    this.listeners.forEach((fn) => fn(...args));
+  }
+}
+
+const injected = new WeakSet();
+
+export type DevToolsMessage =
+  | {
+      type: "updated_swr_cache";
+      payload: {
+        key: string;
+        value: any;
+      };
+    }
+  | {
+      type: "initialized";
+    }
+  | {
+      type: "request_start";
+      payload: {
+        key: string;
+        id: number;
+      };
+    }
+  | {
+      type: "request_success";
+      payload: {
+        key: string;
+        id: number;
+      };
+    }
+  | {
+      type: "request_error";
+      payload: {
+        key: string;
+        id: number;
+      };
+    }
+  | {
+      type: "request_discarded";
+      payload: {
+        key: string;
+        id: number;
+      };
+    };
+
+// TOOD: we have to support more types
+const convertToSerializableObject = (value: any) => {
+  return value instanceof Error ? { message: value.message } : value;
+};
+
+const inject = (cache: Cache) =>
+  injectSWRCache(cache, (key: string, value: any) => {
+    window.postMessage(
+      {
+        type: "updated_swr_cache",
+        payload: {
+          key,
+          value: Object.keys(value).reduce(
+            (acc, cacheKey) => ({
+              ...acc,
+              [cacheKey]: convertToSerializableObject(value[cacheKey]),
+            }),
+            {}
+          ),
+        },
+      },
+      "*"
+    );
+  });
+
+const noop = () => {
+  /* noop */
+};
+const dummyHooks = {
+  useLayoutEffect: noop,
+  useEffect: noop,
+  useRef: <T>(a: T) => ({
+    current: a,
+  }),
+};
+
+export const createSWRDevtools = () => {
+  const events = new EventEmitter();
+
+  const swrdevtools: Middleware = (useSWRNext) => (key, fn, config) => {
+    // use the same React instance with the application
+    const { useLayoutEffect, useEffect, useRef } =
+      typeof window !== "undefined" &&
+      // @ts-expect-error
+      typeof window.__SWR_DEVTOOLS_REACT__ !== "undefined"
```

Comment:
@koba04 Do you feel we could also use a "hook" or "callback" thing to pass down more stuff? like

```js
const { React: SharedReact } = window.__SWR_DEVTOOLS_HOOK__()
const { useLayoutEffect, useEffect, useRef } = SharedReact
```

once we need anything else besides the shared `React`, we can pass more parameters to the return value of the callback. wdyt?

Reply:
Do you have any example for things other than React that we might need @huozhi?

Reply:
Not yet, I just feel it's easier to not break the protocol, for now we only need `React`, but later we might have others. Returning from a hook function can be pretty flexible

Reply:
@huozhi @shuding 
As @shuding mentioned, React is a special case for that reason, so we don't have other things we'd like to pass.
But something might come up in the future. In this case, I can add a new "hook" like `__SWR_DEVTOOLS_HOOK__` and include `React` in the hook, so I feel `__SWR_DEVTOOLS_REACT__` is enough at this time.

Reply:
Yeah that sounds fair. Maybe a renaming to __SWR_DEVTOOLS_XXX__ else not tight to react itself can be nicer. But it's just my personal thought not a blocking at all for any current change 👍 🚀 

If devtool is tighted with version then we can also change it later when it's necessary